File: apps/desktop/src-tauri/src/workspaces.rs
//! Workspace templates and the last open session, kept in the local store so
//! the UI can offer "Restore previous session" and reopen saved workspaces.
//! Only connection targets and layout are stored — never terminal contents.

use serde::{Deserialize, Serialize};
use tauri::State;
use uuid::Uuid;

use crate::error::{DesktopError, Result};
use crate::sessions::OpenTarget;
use crate::state::AppState;

const META_KEY: &str = "desktop.workspaces";
const MAX_TEMPLATES: usize = 200;
const MAX_NAME: usize = 120;

/// Split tree of a saved tab; leaves are connection targets.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum LayoutTemplate {
    Leaf {
        target: OpenTarget,
    },
    Split {
        /// `row` | `column`.
        direction: String,
        ratio: f32,
        first: Box<LayoutTemplate>,
        second: Box<LayoutTemplate>,
    },
}

impl LayoutTemplate {
    fn leaves(&self) -> usize {
        match self {
            Self::Leaf { .. } => 1,
            Self::Split { first, second, .. } => first.leaves() + second.leaves(),
        }
    }

    fn validate(&self) -> Result<()> {
        if let Self::Split {
            direction,
            ratio,
            first,
            second,
        } = self
        {
            if !matches!(direction.as_str(), "row" | "column") {
                return Err(DesktopError::invalid(
                    "split direction must be row or column",
                ));
            }
            if !(0.0..=1.0).contains(ratio) {
                return Err(DesktopError::invalid("split ratio out of range"));
            }
            first.validate()?;
            second.validate()?;
        }
        Ok(())
    }
}

/// A saved workspace: named layout of connections, reopened from the New Tab page.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkspaceTemplate {
    pub id: Uuid,
    pub name: String,
    /// `split` | `list`.
    pub view_mode: String,
    pub layout: LayoutTemplate,
    pub created_at: String,
    pub updated_at: String,
}

/// One tab as it was open when the app last saved its state.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SnapshotTab {
    /// Workspace name; `None` for a plain session tab.
    pub name: Option<String>,
    pub view_mode: String,
    pub template_id: Option<Uuid>,
    pub layout: LayoutTemplate,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionSnapshot {
    pub saved_at: String,
    pub tabs: Vec<SnapshotTab>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
#[serde(default, rename_all = "camelCase")]
pub struct WorkspacesState {
    pub templates: Vec<WorkspaceTemplate>,
    pub last_session: Option<SessionSnapshot>,
}

impl WorkspacesState {
    fn validate(&self) -> Result<()> {
        if self.templates.len() > MAX_TEMPLATES {
            return Err(DesktopError::invalid("too many workspace templates"));
        }
        for t in &self.templates {
            if t.name.trim().is_empty() || t.name.chars().count() > MAX_NAME {
                return Err(DesktopError::invalid(
                    "workspace name must be 1–120 characters",
                ));
            }
            validate_view_mode(&t.view_mode)?;
            t.layout.validate()?;
        }
        if let Some(s) = &self.last_session {
            for t in &s.tabs {
                validate_view_mode(&t.view_mode)?;
                t.layout.validate()?;
                if t.layout.leaves() > 16 {
                    return Err(DesktopError::invalid("too many panes in a saved tab"));
                }
            }
        }
        Ok(())
    }
}

fn validate_view_mode(mode: &str) -> Result<()> {
    if matches!(mode, "split" | "list") {
        Ok(())
    } else {
        Err(DesktopError::invalid("viewMode must be split or list"))
    }
}

pub fn load(state: &AppState) -> Result<WorkspacesState> {
    Ok(match state.store.secret_meta(META_KEY)? {
        Some(json) => serde_json::from_str(&json)?,
        None => WorkspacesState::default(),
    })
}

#[tauri::command]
pub fn workspaces_get(state: State<'_, AppState>) -> Result<WorkspacesState> {
    load(&state)
}

#[tauri::command]
pub fn workspaces_set(
    state: State<'_, AppState>,
    workspaces: WorkspacesState,
) -> Result<WorkspacesState> {
    workspaces.validate()?;
    state
        .store
        .set_secret_meta(META_KEY, &serde_json::to_string(&workspaces)?)?;
    Ok(workspaces)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn leaf(target: OpenTarget) -> LayoutTemplate {
        LayoutTemplate::Leaf { target }
    }

    #[test]
    fn roundtrips_through_json_with_camel_case_keys() {
        let st = WorkspacesState {
            templates: vec![WorkspaceTemplate {
                id: Uuid::nil(),
                name: "Prod".into(),
                view_mode: "split".into(),
                layout: LayoutTemplate::Split {
                    direction: "row".into(),
                    ratio: 0.5,
                    first: Box::new(leaf(OpenTarget::Local)),
                    second: Box::new(leaf(OpenTarget::Quick {
                        address: "example.org".into(),
                        username: Some("root".into()),
                        port: None,
                        protocol: None,
                    })),
                },
                created_at: "2026-01-01T00:00:00Z".into(),
                updated_at: "2026-01-01T00:00:00Z".into(),
            }],
            last_session: Some(SessionSnapshot {
                saved_at: "2026-01-01T00:00:00Z".into(),
                tabs: vec![SnapshotTab {
                    name: None,
                    view_mode: "list".into(),
                    template_id: None,
                    layout: leaf(OpenTarget::Host {
                        host_id: Uuid::nil(),
                        protocol: None,
                    }),
                }],
            }),
        };
        st.validate().unwrap();
        let json = serde_json::to_string(&st).unwrap();
        assert!(json.contains("\"viewMode\""));
        assert!(json.contains("\"lastSession\""));
        assert!(json.contains("\"host_id\""));
        let back: WorkspacesState = serde_json::from_str(&json).unwrap();
        assert_eq!(back, st);
    }

    #[test]
    fn rejects_bad_names_and_layouts() {
        let bad_name = WorkspacesState {
            templates: vec![WorkspaceTemplate {
                id: Uuid::nil(),
                name: "   ".into(),
                view_mode: "split".into(),
                layout: leaf(OpenTarget::Local),
                created_at: String::new(),
                updated_at: String::new(),
            }],
            last_session: None,
        };
        assert!(bad_name.validate().is_err());

        let bad_split = WorkspacesState {
            templates: vec![],
            last_session: Some(SessionSnapshot {
                saved_at: String::new(),
                tabs: vec![SnapshotTab {
                    name: Some("x".into()),
                    view_mode: "split".into(),
                    template_id: None,
                    layout: LayoutTemplate::Split {
                        direction: "diagonal".into(),
                        ratio: 0.5,
                        first: Box::new(leaf(OpenTarget::Local)),
                        second: Box::new(leaf(OpenTarget::Local)),
                    },
                }],
            }),
        };
        assert!(bad_split.validate().is_err());
    }

    #[test]
    fn missing_fields_default() {
        let st: WorkspacesState = serde_json::from_str("{}").unwrap();
        assert!(st.templates.is_empty());
        assert!(st.last_session.is_none());
    }
}

File: apps/desktop/src/terminal/workspaces.ts
// Workspace templates and "restore previous session". Templates are named
// layouts of connections saved from the New Tab page or a tab; the previous
// session is a snapshot of the open tabs, written to the local store whenever
// they change and offered for restore on the next start. Only targets and
// layout are stored — never terminal contents or credentials.

import * as ipc from "@/ipc/commands";
import type {
  LayoutTemplate,
  OpenTarget,
  SessionSnapshot,
  SnapshotTab,
  TabViewMode,
  Uuid,
  WorkspaceTemplate,
} from "@/ipc/types";
import { createStore, useStore } from "@/lib/store";
import {
  openLayout,
  openTerminal,
  renameTab,
  setTabTemplate,
  tabLayoutTemplate,
  terminalStore,
  type TerminalState,
  type TerminalTab,
} from "./store";
import { tr, msg } from "@/i18n";

export interface WorkspacesUiState {
  loaded: boolean;
  templates: WorkspaceTemplate[];
  /** Tabs open when the app last ran, until restored or dismissed. */
  previous: SessionSnapshot | null;
  /** Template whose name is being edited inline (just created or "Rename"). */
  editingId: Uuid | null;
}

export const workspacesStore = createStore<WorkspacesUiState>({
  loaded: false,
  templates: [],
  previous: null,
  editingId: null,
});

export const useWorkspaces = <S>(selector: (s: WorkspacesUiState) => S) =>
  useStore(workspacesStore, selector);

export const DEFAULT_WORKSPACE_NAME = msg("New Workspace");
const SAVE_DEBOUNCE_MS = 400;

const uuid = () => crypto.randomUUID();
const now = () => new Date().toISOString();

function update(fn: (s: WorkspacesUiState) => WorkspacesUiState) {
  workspacesStore.set(fn);
}

// ───────────────────────────── layouts ─────────────────────────────

/** Leaves of a layout in reading order. */
function nameOrDefault(name: string | undefined): string {
  const trimmed = name?.trim() ?? "";
  return trimmed === "" ? tr(DEFAULT_WORKSPACE_NAME) : trimmed;
}

export function layoutTargets(node: LayoutTemplate): OpenTarget[] {
  if (node.kind === "leaf") return [node.target];
  return [...layoutTargets(node.first), ...layoutTargets(node.second)];
}

/** Balanced layout for a list of targets: rows of up to three, stacked. */
export function layoutFromTargets(targets: OpenTarget[]): LayoutTemplate | null {
  const leaves: LayoutTemplate[] = targets.map((target) => ({ kind: "leaf", target }));
  if (leaves.length === 0) return null;
  const rows: LayoutTemplate[] = [];
  for (let i = 0; i < leaves.length; i += 3) rows.push(chain(leaves.slice(i, i + 3), "row"));
  return chain(rows, "column");
}

function chain(nodes: LayoutTemplate[], direction: "row" | "column"): LayoutTemplate {
  const [head, ...rest] = nodes;
  if (!head) throw new Error("empty layout");
  if (rest.length === 0) return head;
  return {
    kind: "split",
    direction,
    ratio: 1 / nodes.length,
    first: head,
    second: chain(rest, direction),
  };
}

export const sameTarget = (a: OpenTarget, b: OpenTarget) => JSON.stringify(a) === JSON.stringify(b);

// ───────────────────────────── snapshot ─────────────────────────────

export function snapshotTabs(s: TerminalState): SnapshotTab[] {
  const out: SnapshotTab[] = [];
  for (const tab of s.tabs) {
    const layout = tabLayoutTemplate(tab, s);
    if (layout) {
      out.push({ name: tab.name, viewMode: tab.viewMode, templateId: tab.templateId, layout });
    }
  }
  return out;
}

export const snapshotConnections = (snap: SessionSnapshot) =>
  snap.tabs.reduce((n, t) => n + layoutTargets(t.layout).length, 0);

// ───────────────────────────── persistence ─────────────────────────────

let saveTimer: ReturnType<typeof setTimeout> | null = null;
let lastSaved = "";
let saving: Promise<void> = Promise.resolve();

function persist(immediate = false) {
  const st = workspacesStore.get();
  if (!st.loaded) return;
  const tabs = snapshotTabs(terminalStore.get());
  const lastSession: SessionSnapshot | null =
    tabs.length > 0 ? { savedAt: now(), tabs } : st.previous;
  const body = { templates: st.templates, lastSession };
  const key = JSON.stringify({ templates: st.templates, tabs: lastSession?.tabs ?? [] });
  if (key === lastSaved) return;
  const write = () => {
    lastSaved = key;
    saving = saving
      .then(() => ipc.workspacesSet(body).then(() => undefined))
      .catch(() => undefined);
  };
  if (saveTimer) clearTimeout(saveTimer);
  if (immediate) {
    saveTimer = null;
    write();
  } else {
    saveTimer = setTimeout(() => {
      saveTimer = null;
      write();
    }, SAVE_DEBOUNCE_MS);
  }
}

let started = false;

/** Load templates and the previous session, then mirror open tabs to the store. */
export function startWorkspaces() {
  if (started) return;
  started = true;
  void ipc
    .workspacesGet()
    .then((st) => {
      const previous = st.lastSession && st.lastSession.tabs.length > 0 ? st.lastSession : null;
      lastSaved = JSON.stringify({ templates: st.templates, tabs: previous?.tabs ?? [] });
      update((w) => ({ ...w, loaded: true, templates: st.templates, previous }));
    })
    .catch(() => update((w) => ({ ...w, loaded: true })))
    .finally(() => persist());
  // Panes are watched too: the shell's cwd / running command live there.
  let seen = terminalStore.get();
  terminalStore.subscribe(() => {
    const s = terminalStore.get();
    if (s.tabs === seen.tabs && s.panes === seen.panes) return;
    seen = s;
    persist();
  });
  window.addEventListener("beforeunload", () => persist(true));
}

// ───────────────────────────── templates ─────────────────────────────

export function templateById(id: Uuid | null, s = workspacesStore.get()) {
  return id ? s.templates.find((t) => t.id === id) : undefined;
}

/** Save a new template and open its name for editing. */
export function createTemplate(
  layout: LayoutTemplate,
  opts: { name?: string; viewMode?: TabViewMode; edit?: boolean } = {},
): WorkspaceTemplate {
  const at = now();
  const tpl: WorkspaceTemplate = {
    id: uuid(),
    name: nameOrDefault(opts.name),
    viewMode: opts.viewMode ?? "split",
    layout,
    createdAt: at,
    updatedAt: at,
  };
  update((s) => ({
    ...s,
    templates: [...s.templates, tpl],
    editingId: opts.edit === false ? s.editingId : tpl.id,
  }));
  persist();
  return tpl;
}

export function renameTemplate(id: Uuid, name: string) {
  const trimmed = name.trim();
  update((s) => ({
    ...s,
    editingId: s.editingId === id ? null : s.editingId,
    templates: trimmed
      ? s.templates.map((t) => (t.id === id ? { ...t, name: trimmed, updatedAt: now() } : t))
      : s.templates,
  }));
  if (trimmed) {
    terminalStore.set((ts) => ({
      ...ts,
      tabs: ts.tabs.map((t) => (t.templateId === id ? { ...t, name: trimmed } : t)),
    }));
  }
  persist();
}

export function deleteTemplate(id: Uuid) {
  update((s) => ({
    ...s,
    editingId: s.editingId === id ? null : s.editingId,
    templates: s.templates.filter((t) => t.id !== id),
  }));
  terminalStore.set((ts) => ({
    ...ts,
    tabs: ts.tabs.map((t) => (t.templateId === id ? { ...t, templateId: null } : t)),
  }));
  persist();
}

export function setEditingTemplate(id: Uuid | null) {
  update((s) => (s.editingId === id ? s : { ...s, editingId: id }));
}

/** Open the template as a new workspace tab connecting every saved target. */
export function openTemplate(id: Uuid, background = false): string | null {
  const tpl = templateById(id);
  if (!tpl) return null;
  return openLayout(tpl.layout, {
    name: tpl.name,
    viewMode: tpl.viewMode,
    templateId: tpl.id,
    background,
  });
}

export interface WorkspaceChoice {
  /** Open workspace tab, or a saved template that is not open right now. */
  kind: "tab" | "template";
  id: string;
  name: string;
}

/** Targets for the "Add to Workspace" submenu: open workspaces, then saved ones. */
export function workspaceChoices(
  tabs: TerminalTab[] = terminalStore.get().tabs,
  templates: WorkspaceTemplate[] = workspacesStore.get().templates,
): WorkspaceChoice[] {
  const open = tabs.filter((t) => t.name !== null);
  const openTemplates = new Set(open.map((t) => t.templateId));
  return [
    ...open.map((t): WorkspaceChoice => ({ kind: "tab", id: t.id, name: t.name ?? "" })),
    ...templates
      .filter((t) => !openTemplates.has(t.id))
      .map((t): WorkspaceChoice => ({ kind: "template", id: t.id, name: t.name })),
  ];
}

/**
 * Add connections to a workspace: `null` starts a new one holding just these
 * targets; a template that is not open is opened first. Alternates split
 * direction so several hosts land in a grid rather than a strip.
 */
export function addToWorkspace(
  choice: WorkspaceChoice | null,
  targets: OpenTarget[],
  background = false,
): string | null {
  if (targets.length === 0) return null;
  if (choice === null) {
    const layout = layoutFromTargets(targets);
    return layout ? openLayout(layout, { name: tr(DEFAULT_WORKSPACE_NAME), background }) : null;
  }
  let tabId: string | null = choice.id;
  if (choice.kind === "template") tabId = openTemplate(choice.id, background);
  if (!tabId) return null;
  let direction: "row" | "column" = "row";
  for (const target of targets) {
    if (openTerminal(target, { intoTab: tabId, direction, background }) === null) break;
    direction = direction === "row" ? "column" : "row";
  }
  return tabId;
}

/**
 * Store the tab's current layout as a template: updates the one it was
 * opened from, otherwise creates a new template named after the tab.
 */
export function saveTabAsTemplate(tabId: string): WorkspaceTemplate | null {
  const s = terminalStore.get();
  const tab = s.tabs.find((t) => t.id === tabId);
  const layout = tab ? tabLayoutTemplate(tab, s) : null;
  if (!tab || !layout) return null;
  const existing = templateById(tab.templateId);
  if (existing) {
    const updated = { ...existing, layout, viewMode: tab.viewMode, updatedAt: now() };
    update((w) => ({
      ...w,
      templates: w.templates.map((t) => (t.id === existing.id ? updated : t)),
    }));
    persist();
    return updated;
  }
  const tpl = createTemplate(layout, {
    name: tab.name ?? undefined,
    viewMode: tab.viewMode,
    edit: false,
  });
  setTabTemplate(tabId, tpl.id);
  if (tab.name === null) renameTab(tabId, tpl.name);
  return tpl;
}

/** The open tab differs from the template it came from ("unsaved changes" dot). */
export function tabDiffersFromTemplate(tab: TerminalTab, s: TerminalState): boolean {
  const tpl = templateById(tab.templateId);
  if (!tpl) return false;
  const layout = tabLayoutTemplate(tab, s);
  return (
    JSON.stringify(layout && layoutTargets(layout)) !== JSON.stringify(layoutTargets(tpl.layout))
  );
}

// ───────────────────────────── previous session ─────────────────────────────

/** Reopen every tab from the previous session; the first one becomes active. */
export function restorePrevious(): number {
  const snap = workspacesStore.get().previous;
  if (!snap) return 0;
  update((s) => ({ ...s, previous: null }));
  let first: string | null = null;
  for (const tab of snap.tabs) {
    const id = openLayout(tab.layout, {
      name: tab.name,
      viewMode: tab.viewMode,
      templateId: templateById(tab.templateId) ? tab.templateId : null,
      background: first !== null,
    });
    first ??= id;
  }
  return snapshotConnections(snap);
}

export function dismissPrevious() {
  update((s) => ({ ...s, previous: null }));
  persist();
}
